src/palette.go

```diff
@@ -0,0 +1,101 @@
+package main
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+)
+
+type Palette map[string]string
+
+const (
+	paletteKeyError          = "palette: requested color %s does not exist in palette of colors %s"
+	paletteRecursiveKeyError = "palette: resolution of color %s returned palette reference %s; recursive references are not supported"
+)
+
+var (
+	paletteColorPrefixes = [...]string{"palette:", "p:"}
+)
+
+// ResolveColor gets a color value from the palette using given colorName.
+// If colorName is not a palette reference, it is returned as is.
+func (p Palette) ResolveColor(colorName string) (string, error) {
+	key, ok := p.asPaletteKey(colorName)
+	// colorName is not a palette key, return it as is
+	if !ok {
+		return colorName, nil
+	}
+
+	color, ok := p[key]
+	if !ok {
+		return "", &PaletteKeyError{Key: key, palette: p}
+	}
+
+	if _, ok = p.asPaletteKey(color); ok {
+		return "", &PaletteRecursiveKeyError{Key: key, Value: color}
+	}
+
+	return color, nil
+}
+
+func (p Palette) asPaletteKey(colorName string) (string, bool) {
+	var (
+		prefix      string
+		prefixFound bool
+	)
+
+	for _, prefix = range paletteColorPrefixes {
+		if strings.HasPrefix(colorName, prefix) {
+			prefixFound = true
+			break
+		}
+	}
+
+	if !prefixFound {
+		return "", false
+	}
+
+	key := strings.ReplaceAll(colorName, prefix, "")
```

Comment:
this can be a simple `strings.TrimPrefix(colorName, prefix)`

Reply:
Correct, changed it.